subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/InputChangeAction.java

```diff
@@ -43,7 +45,7 @@ public void execute(InputFileDetails input) {
         }
 
         File file = input.getFile();
-        if (hasExtension(file, ".java")) {
+        if (hasExtension(file, extension)) {
```

Comment:
This needs to be turned around. For Groovy (i.e. extension = `.groovy`) it is wrong, since `.java` files can also be in the source directory (note: we need test coverage here).
Moreover, for Groovy we have input changes, so we can query `InputChanges` directly for only changes to the sources, no need to use the extension at all.